web_monitoring/diffing_server.py

```diff
@@ -74,16 +74,23 @@ def __init__(self, url, body, headers):
 
 access_control_allow_origin_header = \
     os.environ.get('ACCESS_CONTROL_ALLOW_ORIGIN_HEADER')
-
+ALLOWED_ORIGINS = set()
+if access_control_allow_origin_header:
+    ALLOWED_ORIGINS = set([orig.strip() for orig in
+                           access_control_allow_origin_header.split(',')])
 
 class BaseHandler(tornado.web.RequestHandler):
 
     def set_default_headers(self):
         if access_control_allow_origin_header is not None:
-            self.set_header("Access-Control-Allow-Origin",
-                            access_control_allow_origin_header)
-            self.set_header("Access-Control-Allow-Credentials", "true")
-            self.set_header("Access-Control-Allow-Headers", "x-requested-with")
+            req_origin = self.request.headers.get('Origin')
+            if req_origin and req_origin in ALLOWED_ORIGINS:
+                self.set_header('Access-Control-Allow-Origin', req_origin)
+            else:
+                self.set_header('Access-Control-Allow-Origin',
+                                access_control_allow_origin_header)
```

Comment:
If we know `*` won’t actually work in practice with most good browsers, I don’t think it makes any sense for us to continue supporting it here. We should either reflect the request’s origin or not set this header at all (take your pick).

Reply:
`*` will work in latest FF and Chrome if you don't pass credentials in your CORS request. That's why I wanted to support it.

If you insist, I will only support request's origin, provided its into the allowed origins conf.
Thanks.

Reply:
Ah, I guess that makes some sense as a reason to keep around.

I guess part of me feels like the fact that `*` will only work sometimes is likely to be surprising/confusing for a lot of folks. What if we do this?

```py
req_origin = self.request.headers.get('Origin')
if access_control_allow_origin_header is not None and req_origin is not None:
    if req_origin in ALLOWED_ORIGINS or '*' in ALLOWED_ORIGINS:
        self.set_header('Access-Control-Allow-Origin', req_origin)

    # Set the other headers
```

Is that throwing out an important security feature?

Reply:
If that's OK with you, I've got no problem at all.

Reply:
👍

Reply:
Done.